### DISCUSSION TITLE
How to get the api key working

### DISCUSSION THREAD

**Discussion by BlackViper07** [brianpetro/obsidian-smart-connections#155] [General]
It might be simple, but I am unable of making this plugin work. I made an api key on open ai and entered it in the plugin, but It gives me an error message every time I try to start a chat.

**Comment by brianpetro** [comment]
If it's a new account, the issue is likely rate limits https://github.com/brianpetro/obsidian-smart-connections/discussions/67#discussion-4985289

If you setup billing in your OpenAI account, that should solve it. 

Thanks for trying Smart Connections 🤓
Brian 🌴

**Reply by teatreeman** [reply]
i am also getting this issue, i have gpt premium so i dont think its rate limits? my error says the api key is not working as expected.

**Reply by iamdanielfreitas** [reply]
I have the same problem!

**Reply by zeroliu** [reply]
I also had issues at the beginning but it resolves by itself after setting up the billing and retries a few times. The error message in the developer console can help with debugging.

**Comment by rctheinventor** [comment]
i have premium, i have tried multiple keys, i have tried all the engines.  Can someone suggest some next steps to help me debug this problem?  The plugin doesn't work for me at all right now, and I have tried 3 different keys on my premium account, and I have tried every type of engine, and they all give the same error ... "key is not working as expected".  

The same queries work fine in other web and mobile apps with the same account, and the same keys work in my other test app.  This problem seems to be specific to the Smart Connections plugin, based on my current investigation and debugging.  

Can you please help?

**Reply by brianpetro** [reply]
It may sound silly, but most often this occurs because of a malformed copy & paste. For example, missing or adding a character at the beginning or end of the key.

Also, you may want to make sure you're adding the key to the correct field (i.e. not the "supporter key" input).

I hope that helps.

🌴

**Reply by brianpetro** [reply]
Were you able to make it work @rctheinventor ?

🌴

**Comment by blazingly-fast** [comment]
I have the same problem. Changed the key multiple times, always the same error.

**Reply by brianpetro** [reply]
Hi @blazingly-fast 

If the "test API key" button is returning an error, unfortunately there isn't much that can be done on the Smart Connections side of things besides making sure you copied and pasted the key correctly.

Maybe you could try sending the OpenAI support a message to make sure your account is in good standing with them.

I hope it starts working for you soon!

🌴

**Comment by rctheinventor** [comment]
Hi, I still have the same problem.  I checked for a malformed copy/paste, I deleted and recreated new API keys, I confirmed the status of my Open AI account.  I'm still getting the same error.  

I have a python jupyter notebook on another repo on the same machine (mac mini) that is working ok with the API key, but I'm getting some syntax errors with the new API, so I'm still investigating the notebook errors in parallel, but that runs in a different project/env so I don't think it's related to the Obsidian Smart Connections errors.  I have the latest versions of everything.

UPDATE:  IT WORKS!

I cleaned out everything and started from scratch and it works now! :-)

Thanks!

RC

**Reply by Bail1241** [reply]
I am having the same problem with the API key not working as expected. As I am a premium user, there is no problem with my account and I have created and carefully copied a new API key but still no joy.
I have tried Uninstalling Smart Commections and reinstalling several times but again no joy.
What does rctheinventor above mean by "I cleaned out everything and started from scratch" Is there more to cleaning out than uninstalling?
Any help or comments gratefully received as I really would like Smart Connections to work

David

**Reply by Tomentosa** [reply]
What do you mean 'Cleaned out everything?' Looking to recreate your steps - I'm having the same issue!

**Reply by rctheinventor** [reply]
Hi - sorry, I will be more clear :-) 

1. I uninstalled and then reinstalled the Smart Connections plugin.
2. I deleted all my previous API keys and created a new one called "Obsidian", and added that key into the Smart Connections config.
3. I checked my payment options and account status on https://chat.open.ai - all good there
4. I checked my payment options and account status on https://platform.openai.com/ - there was an outstanding $1.70 on the API charges there and my credit card had expired, so I updated it with the new credit card information.
5. I retested with the new "Obsidian" API key and it worked.

Root cause was 
1. the payment in platform.openai.com, not the payment in chat.open.ai
2. Old API keys 

Once I addressed the root cause, the problem was fixed.  I found it easier to address the root cause by removing all the old keys and creating a new one, but that's just my OCD :-).

**Reply by Tomentosa** [reply]
Ah, I had no idea about the second billing setup for API! Thanks for the quick reply - hopefully this helps other folks too. Successful api key test!

**Reply by rctheinventor** [reply]
:-)

**Reply by J-Louie** [reply]
I ran into this same problem, thought I already did the billing since I'm premium and didn't realize there was a second Billing setup for API's. After setting up the billing for API's,  Smart Connections WORKS!!! Thanks for the help.

**Reply by Bail1241** [reply]
Thank you J -Louie for your comments. I also did not realise there was a second billing setup for APIs. Although I am a premium member myself and this all works fine, I cannot find where to go on the Open AI website to set up the separate API billing. Can you send me the link please? Or perhaps exact instructions on where to go on the website?Sorry to ask such basic questions but I can't get any nearer using Smart Connections until I can set up the separate billing. Thanks again your help.

**Reply by Bail1241** [reply]
I am extremely pleased to announce that Smart Connections is now working for me at last after weeks of problems. As soon as I set up separate API billing all became sweetness and light.  What I would have given to have known that from the beginning! A learning journey indeed……

**Reply by sccardais** [reply]
@Bail1241 

Does this mean that we have to pay for two OpenAI accounts? I have a ChatGPT Pro account that I pay every month. To use Smart Connections, do I need to create another account with OpenAI and pay another $20/month?

Thanks
ScottC
Saluda, NC

**Comment by Bail1241** [comment]
Sent from my iPhoneScottCTwo points 1 yes you have to have the normal OpenAI account and an OpenAI API account which can be set up in the API section of the website. Only with both of these will Smart Connections work 2 you open the API account by paying Say USD 5 and then use Smart Connections until your credit runs out when you must top up,. All working fine for me so far so suggest you give it a try. By copy to Brian After deep frustration with Smart Connections because I had not junderstood the need for the API account I am now a delighted user of Smart Connections which is a truly first class plugin . As a certain British playwright once wrote “all’s well that ends well!”On Dec 21, 2023, at 13:45, sccardais ***@***.***> wrote:﻿
@Bail1241
Does this mean that we have to pay for two OpenAI accounts? I have a ChatGPT Pro account that I pay every month. To use Smart Connections, do I need to create another account with OpenAI and pay another $20/month?
Thanks
ScottC
Saluda, NC

**Reply by sccardais** [reply]
Thank you @Bail1241

So, the API account is "pay as you go." It's not a fixed $20/month. Is that correct?

In your example of USD 5, does that mean that you can set a limit of the expense?

I know the $5 was just an example and that it could be more depending on usage but it seems very reasonable and easy to try.

 Thanks very much.

**Comment by sccardais** [comment]
@brianpetro 

This is a suggestion and not a complaint. 

To make it easier for users to start using Smart Connections, consider adding a clarification in the configuration Settings for Smart Connections. 

Specify that a ChatGPT Pro account alone is insufficient to utilize Smart Connections. Accessing OpenAI via API key requires a separate billing. The good news is that trying it out is both simple and cost-effective, as you can set a low spending limit.

I spent a lot of time trying to figure this out and it looks like others have, too. I created a new key in my ChatGPT OpenAI account that didn't work for Smart Connections and I couldn't figure out why. The OpenAI pages don't explain this at all as far as I can tell. Fortunately, @Bail1241 post set me on the right path.

I think adding a note about this to your configuration page would save people a lot of time.

Thanks for Smart Connections.

**Reply by brianpetro** [reply]
I'll keep this in mind as I'm working on v2. I know it's confused many people so I'll have to add some sort of note.

🌴

**Comment by Bail1241** [comment]
ScottC

USD 5 is the minimum As long as there is some credit I think there is no problem. So you can set 5,10 or 20 or more dollars as your current limit and then re=assess down the line You can also look at your usage and expenditure stats for both accounts on the website.

Good luck!

David

**Comment by bydb** [comment]
After the update my API Key for OpenAI doenst work anymore. Everthing is payed i have a payed openai account

**Reply by brianpetro** [reply]
Hey @bydb 

You can check the developer console for errors.

But it's most likely an issue with your OpenAI account that I can't help with. 

🌴

**Reply by bydb** [reply]
problem solved had tp pay extra for api 🤪

**Comment by aransastarpon** [comment]
I can see this question has been asked but I want to try to be more specific.  I believe that I have properly set up an OpenAI API billing account at this URL https://platform.openai.com/account.  I am getting an API Key error whenever I try to setup this plugin.  Is there some other place to set up an Open AI API account and provision keys.  This is the only place I've been able to find.

**Reply by brianpetro** [reply]
@aransastarpon @bydb it seems like there is a problem with newer API keys. 

I created one about an hour ago to test this issue. The API key worked.

But a little over an hour later, it stopped working. Then I had to revert to my old API key.

I recommend reaching out to OpenAI about this. The more people that reach out to them, the faster it should be solved.

🌴

**Comment by andyjessop** [comment]
I'm getting an error here with a fresh API key (copied via the UI from Anthropic):

```
plugin:smart-connections:5787 TypeError: Cannot read properties of undefined (reading 'tool_choice')
    at ScActions.prepare_request_body (plugin:smart-connections:7540:65)
    at ScChatModel.complete (plugin:smart-connections:5537:132)
    at async ScChatModel.test_api_key (plugin:smart-connections:5781:24)
    at async SmartChatSettings.test_chat_api_key (plugin:smart-connections:6473:22)
```

<img width="679" alt="Screenshot 2024-04-25 at 15 33 48" src="https://github.com/brianpetro/obsidian-smart-connections/assets/1304307/b176ea44-01d1-480a-9b88-67ccd135927c">

**Reply by brianpetro** [reply]
Hey @andyjessop 

Thanks for letting me know.

Try disabling and then re-enabling Smart Connections to see if that changes anything.

🌴
